packages/core/src/browser/shell/application-shell.ts

```diff
@@ -561,9 +561,16 @@ export class ApplicationShell extends Widget {
      * to widgets; these need to be transformed before the layout can be serialized.
      */
     getLayoutData(): ApplicationShell.LayoutData {
+        const pinned: boolean[] = [];
+
+        toArray(this.mainPanel.widgets()).forEach((a, i) => {
+            pinned[i] = a.title.className.indexOf('theia-mod-pinned') >= 0
+        });
```

Comment:
maybe just this:
```typescript
const pinned = toArray(this.mainPanel.widgets()).map(widget => widget.title.className.includes('theia-mod-pinned');
```

Alternatively, since there's some uncertainty below about whether the number of widgets examined here will be the same as the number of widgets eventually restored, you could use something like `WidgetManager.getDescription()` to get a (probably) unique string(ifiable) identifier for each widget, use that as the key in a `Record<string, boolean>` and then match that up with the restored widgets on the other side of the restoration.